proxylib/memcached/text/parser.go

```diff
@@ -0,0 +1,329 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// text memcache protocol parser based on https://github.com/memcached/memcached/blob/master/doc/protocol.txt
+
+package text
+
+import (
+	"bytes"
+	"strconv"
+
+	"github.com/cilium/cilium/pkg/envoy/cilium"
+	"github.com/cilium/cilium/proxylib/memcached/meta"
+	"github.com/cilium/cilium/proxylib/proxylib"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// ParserFactory implements proxylib.ParserFactory
+type ParserFactory struct{}
+
+// Create creates binary memcached parser
+func (p *ParserFactory) Create(connection *proxylib.Connection) proxylib.Parser {
+	log.Infof("ParserFactory: Create: %v", connection)
+	return &Parser{connection: connection, replyQueue: make([]*replyIntent, 0)}
+}
+
+// compile time check for interface implementation
+var _ proxylib.ParserFactory = &ParserFactory{}
+
+// ParserFactoryInstance creates text parser for unified memcached parser
+var ParserFactoryInstance *ParserFactory
+
+// Parser implements proxylib.Parser
+type Parser struct {
+	connection *proxylib.Connection
+
+	requestCount uint32
+	replyCount   uint32
+	replyQueue   []*replyIntent
+
+	//set to true when watch command is observed
+	watching bool
+}
+
+type replyIntent struct {
+	command   []byte
+	requestID uint32
+	denied    bool
+}
+
+var _ proxylib.Parser = &Parser{}
+
+const (
+	CAS_WITH_NOREPLY_FIELDS     = 7
+	STORAGE_WITH_NOREPLY_FIELDS = 6
+	DELETE_WITH_NOREPLY_FIELDS  = 3
+	INCR_WITH_NOREPLY_FIELDS    = 4
+	TOUCH_WITH_NOREPLY_FIELDS   = 4
+)
+
+// OnData parses text memcached data
+func (p *Parser) OnData(reply, endStream bool, dataBuffers [][]byte) (proxylib.OpType, int) {
+	if reply {
+		if p.injectFromQueue() {
+			return proxylib.INJECT, len(DeniedMsg)
+		}
+		if len(dataBuffers) == 0 {
+			return proxylib.NOP, 0
+		}
+	}
+
+	// TODO: don't copy data to new slices
+	data := (bytes.Join(dataBuffers, []byte{}))
+	log.Debugf("Data length: %d", len(data))
+
+	linefeed := bytes.Index(data, []byte("\r\n"))
+	if linefeed < 0 {
+		log.Debugf("Did not receive full first line")
+		if data[len(data)-1] == '\r' {
+			return proxylib.MORE, 1
+		} else {
+			return proxylib.MORE, 2
+		}
+	}
+
+	// TODO: iterate over data without copying it to new slices
+	// Tokenizing in memcached is done by spaces: https://github.com/memcached/memcached/blob/master/memcached.c#L2978
+	tokens := bytes.Fields(data[:linefeed])
+
+	if !reply {
+		meta := meta.MemcacheMeta{
+			Command: string(tokens[0]),
+		}
+		command := tokens[0]
+
+		frameLength := 0
+		noreply := false
+		if p.isCommandRetrieval(command) {
```

Comment:
@nebril the switch case is only to make the code more readable e.g.:
```golang
switch {
        case p.isCommandRetrieval(command):
        // get, gets, gat..
        case p.isCommandStorage(command):
         .....
}
```

Reply:
I see, makes sense, will change it.